serialization/src/derive_support.rs

```diff
@@ -0,0 +1,125 @@
+// Copyright (c) 2022 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://spdx.org/licenses/MIT
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Author(s): L. Kuklinek
+
+//! Implementation tools used by code generated by derive macros
+#![doc(hidden)]
+
+pub use static_assertions as sa;
+
+use crate::Input;
+
+/// Input byte stream with a one-byte lookahead
+pub struct Peekable<'a, I> {
+    init: Option<u8>,
+    inner: &'a mut I,
+}
+
+impl<'a, I: Input> Peekable<'a, I> {
+    /// New peekable input
+    pub fn new(inner: &'a mut I) -> Self {
+        Self { init: None, inner }
+    }
+
+    /// Peek the next byte
+    pub fn peek(&mut self) -> Result<u8, crate::Error> {
+        match self.init {
+            Some(b) => Ok(b),
+            None => {
+                let b = self.inner.read_byte()?;
```

Comment:
I'm confused on what "peek" means. In this context, I thought it means reading a stream without changing the state of the stream. But:

- We're using read_byte(), which consumes a byte
- inner is mutable.

Either "peek" is the wrong word for this or I'm misunderstanding how this works.

Reply:
It refers to how it behaves from the point of view of the user of the `Peekable` type. Peeking may (or may not) advance the `inner` input but `Peekable` still behaves as if it was not consumed, i.e. peeking and reading a byte afterwards gives the same value.